### DISCUSSION TITLE
bytes32 type is too strict

### DISCUSSION THREAD

**Discussion by roninjin10** [wagmi-dev/abitype#77] [General]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Package Version

latest

### TypeScript Version

latest

### Current Behavior

The type for bytes32 only accepts hex numbers of type `0x....`.  Have to awkwardly convert bytes to a hex value.   I think not accepting a string makes sense as it could be a mistake but literal bytes should be valid 
<img width="541" alt="image" src="https://user-images.githubusercontent.com/35039927/215933719-9b0edd4e-fead-4565-8c20-c67dcfe9124d.png">

### Expected Behavior

Should loosen type to accept bytes arrays too

### Link to Minimal Reproducible Example ([TypeScript Playground](https://tsplay.dev/wjQvkW))

https://github.com/ethereum-optimism/optimism/blob/ef8d1454ddd68cc6a42129549595cb0f1d84af87/packages/atst/src/index.ts

### Steps To Reproduce

Pass in any bytes array to a contract method that accepts bytes32

<img width="541" alt="image" src="https://user-images.githubusercontent.com/35039927/215933775-019b4a8c-5352-4280-b143-0cdebcd59f6f.png">

**Comment by Raiden1411** [comment]
You could try extending the config like it's shown here.

![image](https://user-images.githubusercontent.com/67233402/216153939-bc9a2bbc-1136-4edd-bc7c-09a2c7eaccb7.png)

For the wagmi package something like this should work 

```ts
import "abitype";

declare module "wagmi/node_modules/abitype" {
  export interface Config {
    BytesType: Uint8Array | string;
  }
}

```

**Comment by roninjin10** [comment]
Nice!   I still think it should be the default.   It’s odd for literal bytes to not match type for something expecting bytes

**Comment by roninjin10** [comment]
Looks really easy to make it the default I can do the pr

**Comment by Raiden1411** [comment]
Ethers does use something similar [Reference](https://github.com/ethers-io/ethers.js/blob/bf0b468490cb293cd916e4fff06e0909273719e6/packages/bytes/src.ts/index.ts#L10)

Could possibly make sense in having a similar default

**Comment by tmm** [comment]
The default used to be [`string | ArrayLike<number>`](https://github.com/wagmi-dev/abitype/blob/5307b1ec88afae6fd3b69a05777e0fc310db3111/src/config.ts#L32) and I changed it to `` `0x${string}` `` to be stricter. Ideally, the defaults represent the most basic types that still adhere to the ABI spec.

When thinking about defaults, I think about ABI item inputs and outputs. Does `` `0x${string}` | Uint8Array `` make sense? For inputs, they both work since you likely need to do some encoding to get usable data. For outputs though, likely you would decode to a single type `` `0x${string}` `` insead of the union of both values.

For example, if I want to decode `'0x0123456789abcdef000000000000000000000000000000000000000000000000'` with `'bytes8'` as the ABI type, the most basic result type is `` `0x${string}` `` (value: `'0x0123456789abcdef'`) versus converting that to `Uint8Array`.

Think there are two approaches here:
- Keep default `BytesType` the same and consumers can configure ABIType however they like
- Allow for configurable input and output types (something like `BytesInputType`, `BytesOutputType`)

**Reply by Raiden1411** [reply]
I think that option two could possibly make more sense given that [signMessage](https://github.com/wagmi-dev/wagmi/blob/main/packages/core/src/actions/accounts/signMessage.ts) already uses something similar to this and there are other methods that could possibly leverage from this.

**Reply by tmm** [reply]
Agree. Let's go with that. PR welcome!

**Comment by Raiden1411** [comment]
Being worked on #78

**Comment by roninjin10** [comment]
fwiw I disagree with past me here. Consistently using hex strings is quite nice

**Comment by roninjin10** [comment]
I always wonder if there are peformance reasons to use bytes array rather than hex strings though
